File: app/src/main/java/ua/com/programmer/simpleremote/DocumentsListFragment.java
package ua.com.programmer.simpleremote;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

import ua.com.programmer.simpleremote.settings.Constants;
import ua.com.programmer.simpleremote.specialItems.DataBaseItem;
import ua.com.programmer.simpleremote.specialItems.DocumentField;

public class DocumentsListFragment extends Fragment{

    private Context mContext;
    private OnFragmentInteractionListener mListener;
    private SwipeRefreshLayout swipeRefreshLayout;
    private DocumentsAdapter documentsAdapter;

    public DocumentsListFragment() {
        // Required empty public constructor
    }

    static DocumentsListFragment newInstance() {
        return new DocumentsListFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_documents_list, container, false);

        swipeRefreshLayout = fragmentView.findViewById(R.id.documents_swipe);
        swipeRefreshLayout.setOnRefreshListener(this::updateList);

        RecyclerView recyclerView = fragmentView.findViewById(R.id.documents_recycler);
        recyclerView.setHasFixedSize(true);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(mContext);
        recyclerView.setLayoutManager(linearLayoutManager);
        documentsAdapter = new DocumentsAdapter();
        recyclerView.setAdapter(documentsAdapter);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (mListener != null){
                    mListener.onListScrolled(dy);
                }
            }
        });

        return fragmentView;
    }

    @Override
    public void onResume() {
        updateList();
        super.onResume();
    }

    private void updateList(){
        if (!swipeRefreshLayout.isRefreshing()){
            swipeRefreshLayout.setRefreshing(true);
        }
        if(mListener != null){
            mListener.onDataUpdateRequest();
        }
    }

    private void onListItemClick(int position) {
        if (swipeRefreshLayout.isRefreshing()){
            return;
        }
        if (mListener != null) {
            mListener.onFragmentInteraction(documentsAdapter.getListItem(position));
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
        mContext = context;
    }

    void loadListData(ArrayList<DataBaseItem> items){
        documentsAdapter.loadListItems(items);
        swipeRefreshLayout.setRefreshing(false);
    }

    void loadError(String error){
        swipeRefreshLayout.setRefreshing(false);
        Toast.makeText(mContext, error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(DataBaseItem currentListItem);
        void onDataUpdateRequest();
        void onListScrolled(int dy);
    }

    static class DocumentViewHolder extends RecyclerView.ViewHolder{

        TextView tvNumber;
        TextView tvDate;
        TextView tvCompany;
        TextView tvWarehouse;
        TextView tvField1;
        TextView tvField2;
        TextView tvField3;
        TextView tvField4;
        TextView tvSum;
        TextView tvCheckedText;
        TextView tvNotes;
        ImageView icon;

        DocumentViewHolder(View view){
            super(view);
            tvNumber = view.findViewById(R.id.item_number);
            tvDate = view.findViewById(R.id.item_date);
            tvCompany = view.findViewById(R.id.item_company);
            tvWarehouse = view.findViewById(R.id.item_warehouse);
            tvField1 = view.findViewById(R.id.item_field1);
            tvField2 = view.findViewById(R.id.item_field2);
            tvField3 = view.findViewById(R.id.item_field3);
            tvField4 = view.findViewById(R.id.item_field4);
            tvCheckedText = view.findViewById(R.id.item_text_checked);
            tvNotes = view.findViewById(R.id.item_notes);
            tvSum = view.findViewById(R.id.item_sum);
            icon = view.findViewById(R.id.item_icon);
        }

        void setItemInfo(DataBaseItem dataBaseItem){
            tvNumber.setText(dataBaseItem.getString("number"));
            tvDate.setText(dataBaseItem.getString("date"));
            tvCompany.setText(dataBaseItem.getString("company"));
            tvWarehouse.setText(dataBaseItem.getString("warehouse"));
            tvSum.setText(dataBaseItem.getString("sum"));

            DocumentField field1 = new DocumentField(dataBaseItem.getString("field1"));
            tvField1.setText(field1.value);

            DocumentField field2 = new DocumentField(dataBaseItem.getString("field2"));
            if (field2.hasValue()) {
                tvField2.setText(field2.getNamedValue());
                tvField2.setVisibility(View.VISIBLE);
            }else{
                tvField2.setVisibility(View.GONE);
            }

            DocumentField field3 = new DocumentField(dataBaseItem.getString("field3"));
            if (field3.hasValue()) {
                tvField3.setText(field3.getNamedValue());
                tvField3.setVisibility(View.VISIBLE);
            }else {
                tvField3.setVisibility(View.GONE);
            }

            DocumentField field4 = new DocumentField(dataBaseItem.getString("field4"));
            if (field4.hasValue()) {
                tvField4.setText(field4.getNamedValue());
                tvField4.setVisibility(View.VISIBLE);
            }else {
                tvField4.setVisibility(View.GONE);
            }

            String notes = dataBaseItem.getString("notes");
            if (!notes.equals("")) {
                tvNotes.setText(notes);
                tvNotes.setVisibility(View.VISIBLE);
            }else {
                tvNotes.setVisibility(View.GONE);
            }

            if (dataBaseItem.hasValue("checked") && dataBaseItem.getBoolean("checked")) {
                tvCheckedText.setVisibility(View.VISIBLE);
            }else {
                tvCheckedText.setVisibility(View.INVISIBLE);
            }

            int isProcessed = dataBaseItem.getInt("isProcessed");
            int isDeleted = dataBaseItem.getInt("isDeleted");
            if (dataBaseItem.hasValue(Constants.CACHE_GUID)){
                icon.setImageResource(R.drawable.sharp_help_outline_24);
            }else if (isDeleted == 1) {
                icon.setImageResource(R.drawable.twotone_close_24);
            }else if(isProcessed == 1) {
                icon.setImageResource(R.drawable.twotone_check_box_24);
            }else {
                icon.setImageResource(R.drawable.twotone_check_box_outline_blank_24);
            }
        }
    }

    class DocumentsAdapter extends RecyclerView.Adapter<DocumentViewHolder>{

        private final ArrayList<DataBaseItem> listItems = new ArrayList<>();

        void loadListItems(ArrayList<DataBaseItem> values){
            listItems.clear();
            listItems.addAll(values);
            notifyDataSetChanged();
        }

        DataBaseItem getListItem(int position){
            if (position < getItemCount()){
                return listItems.get(position);
            }
            return new DataBaseItem();
        }

        @NonNull
        @Override
        public DocumentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.documents_list_item,parent,false);
            return new DocumentViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull DocumentViewHolder holder, int position) {
            holder.setItemInfo(getListItem(position));
            holder.itemView.setOnClickListener((View v) -> onListItemClick(position));
        }

        @Override
        public int getItemCount() {
            return listItems.size();
        }
    }
}

File: app/src/main/java/ua/com/programmer/simpleremote/deprecated/DocumentsListFragment.kt
package ua.com.programmer.simpleremote.deprecated

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import ua.com.programmer.simpleremote.R
import ua.com.programmer.simpleremote.deprecated.settings.Constants
import ua.com.programmer.simpleremote.deprecated.specialItems.DataBaseItem
import ua.com.programmer.simpleremote.deprecated.specialItems.DocumentField

class DocumentsListFragment : Fragment() {
    private var mContext: Context? = null
    private var mListener: OnFragmentInteractionListener? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null
    private var documentsAdapter: DocumentsAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val fragmentView = inflater.inflate(R.layout.fragment_documents_list, container, false)

        swipeRefreshLayout = fragmentView.findViewById<SwipeRefreshLayout>(R.id.documents_swipe)
        swipeRefreshLayout!!.setOnRefreshListener(SwipeRefreshLayout.OnRefreshListener { this.updateList() })

        val recyclerView = fragmentView.findViewById<RecyclerView>(R.id.documents_recycler)
        //recyclerView.setHasFixedSize(true);
        val linearLayoutManager = LinearLayoutManager(mContext)
        recyclerView.setLayoutManager(linearLayoutManager)
        documentsAdapter = DocumentsAdapter()
        recyclerView.setAdapter(documentsAdapter)

        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (mListener != null) {
                    mListener!!.onListScrolled(dy)
                }
            }
        })

        return fragmentView
    }

    override fun onResume() {
        updateList()
        super.onResume()
    }

    private fun updateList() {
        if (!swipeRefreshLayout!!.isRefreshing) {
            swipeRefreshLayout!!.isRefreshing = true
        }
        if (mListener != null) {
            mListener!!.onDataUpdateRequest()
        }
    }

    private fun onListItemClick(position: Int) {
        if (swipeRefreshLayout!!.isRefreshing) {
            return
        }
        if (mListener != null) {
            mListener!!.onFragmentInteraction(documentsAdapter!!.getListItem(position))
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context as OnFragmentInteractionListener
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
        mContext = context
    }

    fun loadListData(items: java.util.ArrayList<DataBaseItem?>) {
        documentsAdapter!!.loadListItems(items)
        swipeRefreshLayout!!.isRefreshing = false
    }

    fun loadError(error: String?) {
        swipeRefreshLayout!!.isRefreshing = false
        Toast.makeText(mContext, error, Toast.LENGTH_SHORT).show()
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(currentListItem: DataBaseItem?)
        fun onDataUpdateRequest()
        fun onListScrolled(dy: Int)
    }

    internal class DocumentViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var tvNumber: TextView = view.findViewById<TextView>(R.id.item_number)
        var tvDate: TextView = view.findViewById<TextView>(R.id.item_date)
        var tvCompany: TextView = view.findViewById<TextView>(R.id.item_company)
        var tvWarehouse: TextView = view.findViewById<TextView>(R.id.item_warehouse)
        var tvContractor: TextView = view.findViewById<TextView>(R.id.item_contractor)
        var tvField1: TextView = view.findViewById<TextView>(R.id.item_field1)
        var tvField2: TextView = view.findViewById<TextView>(R.id.item_field2)
        var tvField3: TextView = view.findViewById<TextView>(R.id.item_field3)
        var tvField4: TextView = view.findViewById<TextView>(R.id.item_field4)
        var tvSum: TextView = view.findViewById<TextView>(R.id.item_sum)
        var tvCheckedText: TextView = view.findViewById<TextView>(R.id.item_text_checked)
        var tvNotes: TextView = view.findViewById<TextView>(R.id.item_notes)
        var icon: ImageView = view.findViewById<ImageView>(R.id.item_icon)
        var iconRepeat: ImageView = view.findViewById<ImageView>(R.id.icon_repeat)

        fun setItemInfo(dataBaseItem: DataBaseItem) {
            tvNumber.text = dataBaseItem.getString("number")
            tvDate.text = dataBaseItem.getString("date")
            tvCompany.text = dataBaseItem.getString("company")
            tvWarehouse.text = dataBaseItem.getString("warehouse")

            val contractor = dataBaseItem.getString("contractor")
            if (contractor.isEmpty()) {
                tvContractor.visibility = View.GONE
            } else {
                tvContractor.visibility = View.VISIBLE
                tvContractor.text = contractor
            }

            val currency = dataBaseItem.getString("currency")
            var textSum = dataBaseItem.getString("sum")
            if (!currency.isEmpty()) textSum = "$textSum $currency"
            tvSum.text = textSum

            val field1 = DocumentField(dataBaseItem.getString("field1"))
            tvField1.text = field1.value

            val field2 = DocumentField(dataBaseItem.getString("field2"))
            if (field2.hasValue()) {
                tvField2.text = field2.getNamedValue()
                tvField2.visibility = View.VISIBLE
            } else {
                tvField2.visibility = View.GONE
            }

            val field3 = DocumentField(dataBaseItem.getString("field3"))
            if (field3.hasValue()) {
                tvField3.text = field3.getNamedValue()
                tvField3.visibility = View.VISIBLE
            } else {
                tvField3.visibility = View.GONE
            }

            val field4 = DocumentField(dataBaseItem.getString("field4"))
            if (field4.hasValue()) {
                tvField4.text = field4.getNamedValue()
                tvField4.visibility = View.VISIBLE
            } else {
                tvField4.visibility = View.GONE
            }

            val notes = dataBaseItem.getString("notes")
            if (!notes.isEmpty()) {
                tvNotes.text = notes
                tvNotes.visibility = View.VISIBLE
            } else {
                tvNotes.visibility = View.GONE
            }

            if (dataBaseItem.hasValue("checked") && dataBaseItem.getBoolean("checked")) {
                tvCheckedText.visibility = View.VISIBLE
            } else {
                tvCheckedText.visibility = View.INVISIBLE
            }

            val isProcessed = dataBaseItem.getInt("isProcessed")
            val isDeleted = dataBaseItem.getInt("isDeleted")
            if (dataBaseItem.hasValue(Constants.CACHE_GUID)) {
                icon.setImageResource(R.drawable.sharp_help_outline_24)
            } else if (isDeleted == 1) {
                icon.setImageResource(R.drawable.twotone_close_24)
            } else if (isProcessed == 1) {
                icon.setImageResource(R.drawable.twotone_check_box_24)
            } else {
                icon.setImageResource(R.drawable.twotone_check_box_outline_blank_24)
            }

            if (dataBaseItem.getString("repeated") == "1") {
                iconRepeat.visibility = View.VISIBLE
            } else {
                iconRepeat.visibility = View.GONE
            }
        }
    }

    internal inner class DocumentsAdapter : RecyclerView.Adapter<DocumentViewHolder?>() {
        private val listItems = ArrayList<DataBaseItem?>()

        @SuppressLint("NotifyDataSetChanged")
        fun loadListItems(values: java.util.ArrayList<DataBaseItem?>) {
            listItems.clear()
            listItems.addAll(values)
            notifyDataSetChanged()
        }

        fun getListItem(position: Int): DataBaseItem? {
            if (position < itemCount) {
                return listItems[position]
            }
            return DataBaseItem()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DocumentViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.documents_list_item, parent, false)
            return DocumentViewHolder(view)
        }

        override fun onBindViewHolder(holder: DocumentViewHolder, position: Int) {
            holder.setItemInfo(getListItem(position)!!)
            holder.itemView.setOnClickListener(View.OnClickListener { v: View? ->
                onListItemClick(
                    position
                )
            })
        }

        override fun getItemCount(): Int {
            return listItems.size
        }
    }

    companion object {
        fun newInstance(): DocumentsListFragment {
            return DocumentsListFragment()
        }
    }
}